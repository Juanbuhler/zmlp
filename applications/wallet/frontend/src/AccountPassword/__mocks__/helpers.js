/**
 * onSubmit
 */

export const onSubmit = ({ dispatch }) =>
  dispatch({
    currentPassword: 'foo',
    newPassword: 'bar',
    confirmPassword: 'bar',
  })
