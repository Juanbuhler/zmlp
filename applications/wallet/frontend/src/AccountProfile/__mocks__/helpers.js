/**
 * onSubmit
 */

export const onSubmit = ({ dispatch }) =>
  dispatch({
    firstName: 'Jane',
    lastName: 'Doe',
    showForm: false,
    success: true,
    errors: {},
  })
