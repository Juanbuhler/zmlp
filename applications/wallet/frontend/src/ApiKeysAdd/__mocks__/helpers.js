/**
 * onSubmit
 */

export const onSubmit = ({ dispatch }) =>
  dispatch({
    apikey: { permissions: ['ApiKeyManage'], secretKey: 'FooBarSecretKey' },
  })

/**
 * onCopy
 */

let mockOnCopy = () => {}

export const __setMockOnCopy = fn => {
  mockOnCopy = fn
}

export const onCopy = (...args) => {
  mockOnCopy(...args)
}
