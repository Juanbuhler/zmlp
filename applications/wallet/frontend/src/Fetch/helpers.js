import { mutate } from 'swr'
import camelCase from 'camelcase'

export const getQueryString = (params = {}) => {
  const queryString = Object.keys(params)
    .filter((p) => params[p])
    .map((p) => `${p}=${params[p]}`)
    .join('&')

  return queryString ? `?${queryString}` : ''
}

export const getCsrfToken = () => {
  if (typeof document === 'undefined') return ''

  const { csrftoken } = Object.fromEntries(
    document.cookie.split(/; */).map((c) => {
      const [key, ...v] = c.split('=')
      return [key, decodeURIComponent(v.join('='))]
    }),
  )

  return csrftoken
}

export const fetcher = async (url, options = {}) => {
  const csrftoken = getCsrfToken()

  const response = await fetch(url, {
    headers: {
      'Content-Type': 'application/json;charset=UTF-8',
      'X-CSRFToken': csrftoken,
      ...options.headers,
    },
    ...options,
  })

  if ([401, 403].includes(response.status)) {
    mutate('/api/v1/me/', {}, false)
  }

  if (response.status >= 400) throw response

  try {
    const json = await response.json()
    return json
  } catch (error) {
    return response
  }
}

export const getPathname = ({ pathname }) => {
  return pathname
    .replace(/\?(.*)/, '')
    .replace(/^\/[a-z0-9-]{36}/, '/<projectId>')
    .replace(/\/[a-zA-Z0-9-_]{32}$/, '/<assetId>')
    .replace(/\/([a-z-]*)\/[a-z0-9-]{36}/g, (_, category) => {
      const item = category.replace(/s$/, '')
      return `/${category}/<${camelCase(item)}Id>`
    })
}
