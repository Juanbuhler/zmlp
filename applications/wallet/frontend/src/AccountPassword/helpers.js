import { fetcher, getCsrfToken, parseResponse } from '../Fetch/helpers'
import { logout } from '../Authentication/helpers'

export const onSubmit = async ({
  dispatch,
  state: { currentPassword, newPassword, confirmPassword },
}) => {
  dispatch({ isLoading: true, errors: {} })

  try {
    await fetcher(`/api/v1/password/change/`, {
      method: 'POST',
      body: JSON.stringify({
        oldPassword: currentPassword,
        newPassword1: newPassword,
        newPassword2: confirmPassword,
      }),
    })

    dispatch({
      currentPassword: '',
      newPassword: '',
      confirmPassword: '',
      success: true,
      isLoading: false,
      errors: {},
    })
  } catch (response) {
    const errors = await parseResponse({ response })

    dispatch({ success: false, isLoading: false, errors })
  }
}

export const onReset = async ({ setError, email, googleAuth }) => {
  const csrftoken = getCsrfToken()

  try {
    const response = await fetch(`/api/v1/password/reset/`, {
      method: 'POST',
      headers: {
        'Content-Type': 'application/json;charset=UTF-8',
        'X-CSRFToken': csrftoken,
      },
      body: JSON.stringify({
        email,
      }),
    })

    if (response.status >= 400) throw response

    logout({ googleAuth })({
      redirectUrl: '/?action=password-reset-request-success',
      redirectAs: '/',
    })
  } catch (response) {
    const { global } = await parseResponse({ response })

    setError(global)
  }
}
