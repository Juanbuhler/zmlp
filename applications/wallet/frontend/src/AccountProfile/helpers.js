import { fetcher } from '../Fetch/helpers'
import { storeUser } from '../Authentication/helpers'

export const onSubmit = async ({
  dispatch,
  state: { id, firstName, lastName },
}) => {
  try {
    const user = await fetcher(`/api/v1/users/${id}/`, {
      method: 'PATCH',
      body: JSON.stringify({ firstName, lastName }),
    })

    storeUser({ user })

    dispatch({
      firstName: user.firstName,
      lastName: user.lastName,
      errors: {},
    })
  } catch (response) {
    const errors = await response.json()
    const parsedErrors = Object.keys(errors).reduce((acc, errorKey) => {
      acc[errorKey] = errors[errorKey].join(' ')

      return acc
    }, {})

    dispatch({ errors: parsedErrors })
  }
}
