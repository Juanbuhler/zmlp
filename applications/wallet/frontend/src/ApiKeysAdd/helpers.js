import { fetcher } from '../Fetch/helpers'

export const onSubmit = ({ projectId }) => async ({ name, permissions: p }) => {
  const permissions = Object.keys(p).filter(key => p[key])

  await fetcher(`/api/v1/projects/${projectId}/apikeys/`, {
    method: 'POST',
    body: JSON.stringify({ name, permissions }),
  })
}
