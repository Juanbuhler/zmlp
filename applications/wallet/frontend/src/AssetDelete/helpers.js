import { cache } from 'swr'
import Router from 'next/router'

import { fetcher } from '../Fetch/helpers'
import { formatUrl } from '../Filters/helpers'

export const onDelete = async ({
  projectId,
  assetId,
  query,
  setIsLoading,
  setError,
}) => {
  try {
    setIsLoading(true)

    await fetcher(`/api/v1/projects/${projectId}/assets/${assetId}/`, {
      method: 'DELETE',
    })

    cache.clear()

    return Router.push(
      {
        pathname: '/[projectId]/visualizer',
        query: { query, id: '', action: 'delete-asset-success' },
      },
      `/${projectId}/visualizer${formatUrl({
        id: '',
        query,
        action: 'delete-asset-success',
      })}`,
    )
  } catch (error) {
    setIsLoading(false)
    return setError('There was an error. Please try again.')
  }
}
