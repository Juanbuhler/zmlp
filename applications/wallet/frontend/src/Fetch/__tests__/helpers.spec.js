import { fetcher, revalidate, getQueryString, getPathname } from '../helpers'

describe('<Fetch /> helpers', () => {
  describe('fetcher()', () => {
    it('should fetch data', async () => {
      fetch.mockResponseOnce(JSON.stringify({ id: 12345 }))

      const data = await fetcher('/url')

      expect(data).toEqual({ id: 12345 })
    })

    it('should return the raw response in case of error', async () => {
      fetch.mockResponseOnce(null, { status: 500 })

      try {
        await fetcher('/url')
      } catch (response) {
        expect(response.status).toBe(500)

        expect(response.statusText).toBe('Internal Server Error')

        expect(response).toMatchSnapshot()
      }
    })

    it('should return the raw response if its not a json', async () => {
      fetch.mockResponseOnce(null, { status: 200 })

      try {
        await fetcher('/url')
      } catch (response) {
        expect(response.status).toBe(200)

        expect(response.statusText).toBe('Ok')

        expect(response).toMatchSnapshot()
      }
    })

    it('should logout the user', async () => {
      const mockMutate = jest.fn()

      require('swr').__setMockMutateFn(mockMutate)

      fetch.mockResponseOnce(null, { status: 401 })

      try {
        await fetcher('/url')
      } catch (response) {
        expect(response.status).toBe(401)

        expect(response.statusText).toBe('Unauthorized')

        expect(mockMutate).toHaveBeenCalledWith({})
      }
    })
  })

  describe('revalidate()', () => {
    it('should fetch data', async () => {
      await revalidate({ key: '/url' })

      expect(fetch.mock.calls[0][0]).toEqual('/url')
    })
  })

  describe('getQueryString()', () => {
    it('should return an empty string with no query params', () => {
      expect(getQueryString()).toEqual('')
    })
  })

  describe('getPathname()', () => {
    it('should return a pathname', () => {
      expect(getPathname({ pathname: '/' })).toEqual('/')
    })

    it('should strip query params', () => {
      expect(getPathname({ pathname: '/?foo=bar' })).toEqual('/')
    })

    it('should strip the projectId', () => {
      expect(
        getPathname({ pathname: '/a0952c03-cc04-461c-a367-9ffae8c4199a' }),
      ).toEqual('/<projectId>')
    })

    it('should camelCase', () => {
      expect(
        getPathname({
          pathname:
            '/a0952c03-cc04-461c-a367-9ffae8c4199a/data-sources/81ca313d-dc65-1391-9fe9-1aeeeaea6f33/edit',
        }),
      ).toEqual('/<projectId>/data-sources/<dataSourceId>/edit')
    })

    it('should strip recursively', () => {
      expect(
        getPathname({
          pathname:
            '/a0952c03-cc04-461c-a367-9ffae8c4199a/jobs/bc8e8b24-7aa2-1f49-a1c6-420403fdacd8/tasks/bc8e92b5-7aa2-1f49-a1c6-420403fdacd8/errors/91325a65-cf73-17a9-bd8f-76ee1900ec47',
        }),
      ).toEqual('/<projectId>/jobs/<jobId>/tasks/<taskId>/errors/<errorId>')
    })

    it('should strip the assetId', () => {
      expect(
        getPathname({
          pathname:
            '/a0952c03-cc04-461c-a367-9ffae8c4199a/visualizer/2SdbPKHfNc0CPKBRAyA_V3oM5Oenpt3B',
        }),
      ).toEqual('/<projectId>/visualizer/<assetId>')
    })
  })
})
