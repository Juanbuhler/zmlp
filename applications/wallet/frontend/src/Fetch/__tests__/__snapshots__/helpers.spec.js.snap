// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Fetch /> helpers fetcher() should return the raw response if its neither text nor json 1`] = `
Response {
  "size": 0,
  "timeout": 0,
  Symbol(Body internals): Object {
    "body": null,
    "disturbed": true,
    "error": null,
  },
  Symbol(Response internals): Object {
    "counter": undefined,
    "headers": Headers {
      Symbol(map): Object {
        "content-type": Array [
          "image/jpeg",
        ],
      },
    },
    "status": 200,
    "statusText": "OK",
    "url": undefined,
  },
}
`;

exports[`<Fetch /> helpers fetcher() should return the raw response in case of error 1`] = `
Response {
  "size": 0,
  "timeout": 0,
  Symbol(Body internals): Object {
    "body": null,
    "disturbed": false,
    "error": null,
  },
  Symbol(Response internals): Object {
    "counter": undefined,
    "headers": Headers {
      Symbol(map): Object {
        "content-type": Array [
          "application/json",
        ],
      },
    },
    "status": 500,
    "statusText": "Internal Server Error",
    "url": undefined,
  },
}
`;
